cli/command/service/update.go

```diff
@@ -588,26 +593,58 @@ func portConfigToString(portConfig *swarm.PortConfig) string {
 	if protocol == "" {
 		protocol = "tcp"
 	}
-	return fmt.Sprintf("%v/%s", portConfig.PublishedPort, protocol)
+
+	mode := portConfig.PublishMode
+	if mode == "" {
+		mode = "ingress"
+	}
+
+	return fmt.Sprintf("%v:%v/%s/%s", portConfig.PublishedPort, protocol, mode)
 }
 
 func updatePorts(flags *pflag.FlagSet, portConfig *[]swarm.PortConfig) error {
 	// The key of the map is `port/protocol`, e.g., `80/tcp`
 	portSet := map[string]swarm.PortConfig{}
 	// Check to see if there are any conflict in flags.
-	if flags.Changed(flagPublishAdd) {
-		values := flags.Lookup(flagPublishAdd).Value.(*opts.ListOpts).GetAll()
-		ports, portBindings, _ := nat.ParsePortSpecs(values)
-
-		for port := range ports {
-			newConfigs := convertPortToPortConfig(port, portBindings)
-			for _, entry := range newConfigs {
-				if v, ok := portSet[portConfigToString(&entry)]; ok && v != entry {
-					return fmt.Errorf("conflicting port mapping between %v:%v/%s and %v:%v/%s", entry.PublishedPort, entry.TargetPort, entry.Protocol, v.PublishedPort, v.TargetPort, v.Protocol)
-				}
-				portSet[portConfigToString(&entry)] = entry
+	if flags.Changed(flagPublishAdd) || flags.Changed(flagPortAdd) {
+		var (
+			newConfigs []swarm.PortConfig
+			values     []string
+		)
+
+		for _, flag := range []string{flagPublishAdd, flagPortAdd} {
+			values = append(values, flags.Lookup(flag).Value.(*opts.ListOpts).GetAll()...)
+		}
+
+		for _, value := range values {
+			if strings.Index(value, "=") != -1 {
+				// We can ignore errors because the format was already validated by ValidatePort
+				portKVPairs, _ := parsePortCSV(value)
+				tPort, _ := strconv.ParseUint(portKVPairs[portOptTargetPort], 10, 16)
+				pPort, _ := strconv.ParseUint(portKVPairs[portOptPublishedPort], 10, 16)
+				newConfigs = append(newConfigs, swarm.PortConfig{
+					Protocol:      swarm.PortConfigProtocol(portKVPairs[portOptProtocol]),
+					TargetPort:    uint32(tPort),
+					PublishedPort: uint32(pPort),
+					PublishMode:   swarm.PortConfigPublishMode(portKVPairs[portOptMode]),
+				})
+
+				continue
+			}
```

Comment:
This code will be much cleaner when we have a `PortOpt` type (as requested above).